src/DomainNetMap.cpp

```diff
@@ -0,0 +1,344 @@
+        }
+
+        if (nnet_json.is_null() || !nnet_json.is_object()) {
+            throw Exception("DomainNetMapImp::" + std::string(__func__) +
+                            ": Neural net file format is incorrect: object expected.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        // make sure that there no unexpected keys in the json
+        for (const auto &key : nnet_json.object_items()) {
+            if (std::find(M_EXPECTED_KEYS.begin(), M_EXPECTED_KEYS.end(), key.first)
+                == M_EXPECTED_KEYS.end()) {
+                throw Exception("DomainNetMapImp::" + std::string(__func__) + 
+                                ": Unexpected key in neural net json: " + key.first,
+                                GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+        }
+
+        // will check key exists, it's an array, and is not empty
+        if (nnet_json["layers"].array_items().size() == 0) {
+            throw Exception("DomainNetMapImp::" + std::string(__func__) + 
+                            ": Neural net must contain valid json and must have "
+                            "a key \"layers\" whose value is a non-empty array.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+
+        const auto nnet_properties = nnet_json.object_items();
+        const auto signal_it = nnet_properties.find("signal_inputs");
+        const auto delta_it = nnet_properties.find("delta_inputs");
+        size_t signal_inputs_size = 0;
+        size_t delta_inputs_size = 0;
+
+        if (signal_it != nnet_properties.end()) {
+            if (!signal_it->second.is_array()) {
+                throw Exception("DomainNetMapImp::" + std::string(__func__) +
+                                 ": Neural net \"signal_inputs\" must be an array.",
+                                 GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+            signal_inputs_size = signal_it->second.array_items().size();
+        }
+
+        if (delta_it != nnet_properties.end()) {
+            if (!delta_it->second.is_array()) {
+                throw Exception("DomainNetMapImp::" + std::string(__func__) +
+                                ": Neural net \"delta_inputs\" must be an array.",
+                                 GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+            }
+            delta_inputs_size = delta_it->second.array_items().size();
+        }
+
+        if (signal_inputs_size == 0 && delta_inputs_size == 0) {
+            throw Exception("DomainNetMapImp::" + std::string(__func__) + 
+                            ": Neural net json must contain at least one of "
+                            "\"signal_inputs\" and \"delta_inputs\" whose "
+                            "value is a non-empty array.",
+                            GEOPM_ERROR_INVALID, __FILE__, __LINE__);
+        }
+        
+        std::vector<std::shared_ptr<DenseLayer>> layers;
```

Comment:
nit
```suggestion
        std::vector<std::shared_ptr<DenseLayer> > layers;
```

Reply:
Q: is this for stylistic reasons? I know in earlier versions of C++ this was required, but in newer versions it is not.

Reply:
Purely stylistic since we're on c++17.  We do this else ware in the codebase, and I'd prefer we're consistent.